core/shared/src/main/scala/zio/ZIOApp.scala

```diff
@@ -18,31 +18,37 @@ package zio
 import zio.internal._
 
 /**
- * The entry point for a ZIO application.
- *
- * {{{
- * import zio.ZIOApp
- * import zio.Console._
- *
- * object MyApp extends ZIOApp {
- *
- *   def run =
- *     for {
- *       _ <- printLine("Hello! What is your name?")
- *       n <- readLine
- *       _ <- printLine("Hello, " + n + ", good to meet you!")
- *     } yield ()
- * }
- * }}}
+ * An entry point for a ZIO application that allows sharing layers between
+ * applications. For a simpler version that uses the default ZIO environment
+ * see `ZIOAppDefault`.
  */
 trait ZIOApp { self =>
   @volatile private var shuttingDown = false
 
+  implicit def tag: Tag[Environment]
+
+  type Environment <: Has[_]
+
+  /**
+   * A layer that manages the acquisition and release of services necessary for
+   * the application to run.
+   */
+  def layer: ZLayer[ZEnv with Has[ZIOAppArgs], Any, Environment]
```

Comment:
The `Has[ZIOAppArgs]` is necessary since those come from the `main` function, implemented by `ZIOApp`. One could argue either way on supplying `ZEnv` to `layer`. If we supply it, it has to be the real one, if it has to be the real one, then we can't provide a test helper for ZIO Test (although we could work around this in other ways).

Reply:
I changed the implementation in `ZIOSpec` to override `layer` with a new `final` definition and expose a new `testLayer` method that uses the live environment to build a `TestEnvironment`. We could definitely remove and people could provide `ZEnv` themselves if they needed it.